### PR TITLE
feat: Built-in Jetstream source implementation (Closes #1695)

### LINKED ISSUE #1695: inbuilt JetStream source
# Summary

For async processing, a user-defined source (say, a custom HTTP server) could write to JetStream, and Numaflow can reliably process the data. This should be a relatively straightforward pattern since we use JetStream as the ISB. The JetStream source can be an inbuilt source since we already have code for reading and writing to the stream.

# Use Cases

Ability to insert data into the Numaflow pipeline from a decoupled user-defined source.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented, please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by BulkBeing**
Closes #1695

### CODE REVIEW

**Comment by yhl25** [comment]
run `make codegen`?

**Comment by yhl25** [comment]
ignore` kafka-e2e` and` idle-source-e2e`, its failing because of https://github.com/numaproj/numaflow/issues/1724

**Comment by BulkBeing** [comment]
On my laptop, `make codegen` fails. Execution with trace enabled:
```sh
➜  make codegen
./hack/generate-proto.sh
+ set -o errexit
+ set -o nounset
+ set -o pipefail
++ dirname ./hack/generate-proto.sh
+ source ./hack/library.sh
+++ git rev-parse --show-toplevel
++ readonly REPO_ROOT=/Users/spr01/Documents/GIT/numaflow
++ REPO_ROOT=/Users/spr01/Documents/GIT/numaflow
+ header 'generating proto files'
++ echo generating proto files
++ tr a-z A-Z
+ local 'upper=GENERATING PROTO FILES'
+ make_banner + 'GENERATING PROTO FILES'
+ local 'msg=++++ GENERATING PROTO FILES ++++'
+ local border=++++++++++++++++++++++++++++++++
+ echo -e '++++++++++++++++++++++++++++++++\n++++ GENERATING PROTO FILES ++++\n++++++++++++++++++++++++++++++++'
++++++++++++++++++++++++++++++++
++++ GENERATING PROTO FILES ++++
++++++++++++++++++++++++++++++++
+ ensure_protobuf
++ command -v protoc
+ '[' /opt/homebrew/bin/protoc = '' ']'
+ ensure_vendor
+ go mod vendor
++ command -v protoc-gen-gogo
+ '[' /Users/spr01/go/bin/protoc-gen-gogo = '' ']'
++ command -v protoc-gen-gogofast
+ '[' /Users/spr01/go/bin/protoc-gen-gogofast = '' ']'
++ command -v gogoproto
+ '[' '' = '' ']'
+ go install -mod=vendor ./vendor/github.com/gogo/protobuf/gogoproto
++ command -v protoc-gen-grpc-gateway
+ '[' /Users/spr01/go/bin/protoc-gen-grpc-gateway = '' ']'
++ command -v goimports
+ '[' '' = '' ']'
+ export GO111MODULE=off
+ GO111MODULE=off
+ go get golang.org/x/tools/cmd/goimports
go: modules disabled by GO111MODULE=off; see 'go help modules'
make: *** [codegen] Error 1
```
Moved `golang.org/x/tools/cmd/goimports` to `tools.go` to fix this.

**Comment by vigith** [comment]
> Moved `golang.org/x/tools/cmd/goimports` to `tools.go` to fix this.

Read this thread https://intuit-teams.slack.com/archives/C02P1NKMBHQ/p1712074641682409?thread_ts=1712010787.475059&cid=C02P1NKMBHQ

File: hack/generate-proto.sh [hunk 1]
```diff
@@ -27,9 +27,7 @@ if [ "`command -v protoc-gen-grpc-gateway`" = "" ]; then
 fi
 
 if [ "`command -v goimports`" = "" ]; then
-  export GO111MODULE="off"
-  go get golang.org/x/tools/cmd/goimports
-  export GO111MODULE="on"
+  go install -mod=vendor ./vendor/golang.org/x/tools/cmd/goimports
```
Comment (vigith) [inline hack/generate-proto.sh:30]: why change this?

File: pkg/sources/jetstream/jetstream.go [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+				},
+			},
+			ReadOffset: newOffset(msg, metadata, vertexInstance.Replica),
+		}
+		n.messages <- m
+	}
+
+	consumeErrHandler := func(consumeCtx jetstreamlib.ConsumeContext, err error) {
+		if errors.Is(err, jetstreamlib.ErrNoMessages) || errors.Is(err, jetstreamlib.ErrNoHeartbeat) {
+			n.logger.Infow("Ignoring the error", zap.Error(err))
```
Comment (vigith) [inline pkg/sources/jetstream/jetstream.go:181]: no need to log `ErrNoMessage`.. please log `ErrNoHearbeat` as warn.

**Review by vigith — COMMENTED** [review]
* [`InProgress` ack](https://docs.nats.io/using-nats/developer/develop_jetstream#acknowledging-messages) should be implemented
* NIT: please add comments
* call `Stop` on consumer

File: hack/generate-proto.sh (hunk 1 above)
Comment (BulkBeing) [reply hack/generate-proto.sh:30]: I already had `GOPATH/bin` in my `PATH`. In my case, I'm running `make codegen` for the first time on this laptop, and I didn't had `goimport` installed.
The problem seem to be a change in go get in 1.22 https://go.dev/doc/go1.22#go-command.
> go get is no longer supported outside of a module in the legacy GOPATH mode (that is, with GO111MODULE=off). Other build commands, such as go build and go test, will continue to work indefinitely for legacy GOPATH programs.

Also, the current change makes the installation method same as that of other Go tools in this project.

**Review by vigith — APPROVED** [review]
LGTM

File: pkg/sources/jetstream/jetstream.go [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+		latestErr:     latestErr,
+	}
+	for _, o := range opts {
+		if err := o(n); err != nil {
+			return nil, err
+		}
+	}
+	if n.logger == nil {
+		n.logger = logging.NewLogger()
+	}
```
Comment (yhl25) [inline pkg/sources/jetstream/jetstream.go:78]: `logging.FromContext(ctx)`?

**Review by yhl25 — APPROVED** [review]

**Review by whynowy — APPROVED** [review]
LGTM, thanks for submitting the PR!

A follow-up doc for JetStream source is needed.

**Comment by whynowy** [comment]
@BulkBeing - please also enable `jetstream-e2e` at https://github.com/numaproj/numaflow/blob/main/.github/workflows/ci.yaml#L142.

**Comment by BulkBeing** [comment]
> @BulkBeing - please also enable `jetstream-e2e` at https://github.com/numaproj/numaflow/blob/main/.github/workflows/ci.yaml#L142.

Added

**Pull request merged**